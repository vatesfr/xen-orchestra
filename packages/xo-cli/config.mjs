import { ensureDir as mkdirp } from 'fs-extra'
import { readFile, writeFile } from 'fs/promises'
import { xdgConfig } from 'xdg-basedir'
import lodashGet from 'lodash/get.js'
import lodashUnset from 'lodash/unset.js'

// ===================================================================

const configPath = xdgConfig + '/xo-cli'
const configFile = configPath + '/config.json'

// ===================================================================

export function load() {
  return readFile(configFile)
    .then(JSON.parse)
    .catch(function () {
      return {}
    })
}

export function get(path) {
  return load().then(function (config) {
    return lodashGet(config, path)
  })
}

export function save(config) {
  return mkdirp(configPath).then(function () {
    return writeFile(configFile, JSON.stringify(config))
  })
}

export function set(data) {
  return load().then(function (config) {
    return save(Object.assign(config, data))
  })
}

export function unset(paths) {
  return load().then(function (config) {
    ;[].concat(paths).forEach(function (path) {
      lodashUnset(config, path)
    })
    return save(config)
  })
}
