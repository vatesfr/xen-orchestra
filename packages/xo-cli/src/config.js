const { readFile, writeFile } = require('fs/promises')

const get = require('lodash/get')
const mkdirp = require('fs-extra').ensureDir
const unset = require('lodash/unset')
const xdgBasedir = require('xdg-basedir')

// ===================================================================

const configPath = xdgBasedir.config + '/xo-cli'
const configFile = configPath + '/config.json'

// ===================================================================

const load = (exports.load = function () {
  return readFile(configFile)
    .then(JSON.parse)
    .catch(function () {
      return {}
    })
})

exports.get = function (path) {
  return load().then(function (config) {
    return get(config, path)
  })
}

const save = (exports.save = function (config) {
  return mkdirp(configPath).then(function () {
    return writeFile(configFile, JSON.stringify(config))
  })
})

exports.set = function (data) {
  return load().then(function (config) {
    return save(Object.assign(config, data))
  })
}

exports.unset = function (paths) {
  return load().then(function (config) {
    ;[].concat(paths).forEach(function (path) {
      unset(config, path)
    })
    return save(config)
  })
}
