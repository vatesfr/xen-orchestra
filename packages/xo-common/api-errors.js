'use strict'

const { BaseError } = require('make-error')
const { iteratee } = require('lodash')

class XoError extends BaseError {
  constructor({ code, message, data }) {
    super(message)
    this.code = code
    this.data = data
  }

  toJsonRpcError() {
    return {
      message: this.message,
      code: this.code,
      data: this.data,
    }
  }
}

const create = (code, getProps) => {
  const factory = (...args) => {
    const props = getProps(...args)
    props.code = code
    throw new XoError(props)
  }
  factory.is = (error, predicate) => error.code === code && (predicate === undefined || iteratee(predicate)(error.data))

  return factory
}

// =============================================================================

exports.notImplemented = create(0, () => ({
  message: 'not implemented',
}))

exports.noSuchObject = create(1, (id, type) => ({
  data: { id, type },
  message: `no such ${type || 'object'} ${id}`,
}))

exports.unauthorized = create(2, (permission, objectId, objectType) => ({
  data: {
    permission,
    object: {
      id: objectId,
      type: objectType,
    },
  },
  message: 'not enough permissions',
}))

exports.invalidCredentials = create(3, () => ({
  message: 'invalid credentials',
}))

// Deprecated alreadyAuthenticated (4)

exports.forbiddenOperation = create(5, (operation, reason) => ({
  data: { operation, reason },
  message: `forbidden operation: ${operation}`,
}))

// Deprecated GenericError (6)

exports.noHostsAvailable = create(7, () => ({
  message: 'no hosts available',
}))

exports.authenticationFailed = create(8, () => ({
  message: 'authentication failed',
}))

exports.serverUnreachable = create(9, objectId => ({
  data: {
    objectId,
  },
  message: 'server unreachable',
}))

exports.invalidParameters = create(10, (message, errors) => {
  if (Array.isArray(message)) {
    errors = message
    message = undefined
  }

  return {
    data: { errors },
    message: message || 'invalid parameters',
  }
})

exports.vmMissingPvDrivers = create(11, ({ vm }) => ({
  data: {
    objectId: vm,
  },
  message: 'missing PV drivers',
}))

exports.vmIsTemplate = create(12, ({ vm }) => ({
  data: {
    objectId: vm,
  },
  message: 'VM is a template',
}))

exports.vmBadPowerState = create(13, ({ vm, expected, actual }) => ({
  data: {
    objectId: vm,
    expected,
    actual,
  },
  message: `VM state is ${actual} but should be ${expected}`,
}))

exports.vmLacksFeature = create(14, ({ vm, feature }) => ({
  data: {
    objectId: vm,
    feature,
  },
  message: `VM lacks feature ${feature || ''}`,
}))

exports.notSupportedDuringUpgrade = create(15, () => ({
  message: 'not supported during upgrade',
}))

exports.objectAlreadyExists = create(16, ({ objectId, objectType }) => ({
  data: {
    objectId,
    objectType,
  },
  message: `${objectType || 'object'} already exists`,
}))

exports.vdiInUse = create(17, ({ vdi, operation }) => ({
  data: {
    objectId: vdi,
    operation,
  },
  message: 'VDI in use',
}))

exports.hostOffline = create(18, ({ host }) => ({
  data: {
    objectId: host,
  },
  message: 'host offline',
}))

exports.operationBlocked = create(19, ({ objectId, code }) => ({
  data: {
    objectId,
    code,
  },
  message: 'operation blocked',
}))

exports.patchPrecheckFailed = create(20, ({ errorType, patch }) => ({
  data: {
    objectId: patch,
    errorType,
  },
  message: `patch precheck failed: ${errorType}`,
}))

exports.operationFailed = create(21, ({ objectId, code }) => ({
  data: {
    objectId,
    code,
  },
  message: 'operation failed',
}))

exports.missingAuditRecord = create(22, ({ id, nValid }) => ({
  data: {
    id,
    nValid,
  },
  message: 'missing record',
}))

exports.alteredAuditRecord = create(23, ({ id, record, nValid }) => ({
  data: {
    id,
    record,
    nValid,
  },
  message: 'altered record',
}))

exports.notEnoughResources = create(24, data => ({
  data, // [{ resourceSet, resourceType, available, requested }]
  message: 'not enough resources in resource set',
}))

exports.incorrectState = create(25, ({ actual, expected, object, property }) => ({
  data: {
    actual,
    expected,
    object,
    property,
  },
  message: 'incorrect state',
}))

exports.featureUnauthorized = create(26, ({ featureCode, currentPlan, minPlan }) => ({
  data: {
    featureCode,
    currentPlan,
    minPlan,
  },
  message: 'feature Unauthorized',
}))
