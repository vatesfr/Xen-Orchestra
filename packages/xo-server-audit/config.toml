# List of skipped methods
blockedList = [
  'acl.get',
  'acl.getCurrentPermissions',
  'audit.checkIntegrity',
  'audit.generateFingerprint',
  'audit.getRecords',
  'backupNg.getAllJobs',
  'backupNg.getAllLogs',
  'cloud.getResourceCatalog',
  'cloudConfig.getAll',
  'group.getAll',
  'host.stats',
  'ipPool.getAll',
  'job.getAll',
  'log.get',
  'metadataBackup.getAllJobs',
  'plugin.get',
  'pool.listMissingPatches',
  'proxy.getAll',
  'remote.getAll',
  'remote.getAllInfo',
  'resourceSet.getAll',
  'role.getAll',
  'schedule.getAll',
  'server.getAll',
  'session.getUser',
  'sr.getUnhealthyVdiChainsLength',
  'sr.stats',
  'system.getMethodsInfo',
  'system.getServerTimezone',
  'system.getServerVersion',
  'user.getAll',
  'vm.stats',
  'xo.getAllObjects',
  'xoa.supportTunnel.getState',
  'xosan.checkSrCurrentState',
  'xosan.getVolumeInfo',
]
