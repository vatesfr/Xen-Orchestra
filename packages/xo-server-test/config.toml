# Vendor config: DO NOT TOUCH!
#
# See sample.config.toml to override.

# The clone of a template can exceeds the jest timeout if it's used by other tests to create a VM
cloneTempVmTimeout = '1 minute'

# After some executions we saw that `deleteTempResources` takes around `21s`.
# Then, we chose a large timeout to be sure that all resources created by `xo-server-test`
# will be deleted
deleteTempResourcesTimeout = '2 minutes'
