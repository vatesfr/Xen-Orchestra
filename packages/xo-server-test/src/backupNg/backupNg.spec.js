/* eslint-env jest */

import { forOwn } from 'lodash'
import { noSuchObject } from 'xo-common/api-errors'

import config from '../_config'
import randomId from '../_randomId'
import xo from '../_xoConnection'

const DEFAULT_SCHEDULE = {
  name: 'scheduleTest',
  cron: '0 * * * * *',
}

const validateRootTask = (log, props) =>
  expect(log).toMatchSnapshot({
    end: expect.any(Number),
    id: expect.any(String),
    jobId: expect.any(String),
    scheduleId: expect.any(String),
    start: expect.any(Number),
    ...props,
  })

const validateVmTask = (task, vmId, props = {}) => {
  expect(task).toMatchSnapshot({
    data: {
      id: expect.any(String),
    },
    end: expect.any(Number),
    id: expect.any(String),
    message: expect.any(String),
    start: expect.any(Number),
    ...props,
  })
  expect(task.data.id).toBe(vmId)
}

const validateSnapshotTask = (task, props) =>
  expect(task).toMatchSnapshot({
    end: expect.any(Number),
    id: expect.any(String),
    result: expect.any(String),
    start: expect.any(Number),
    ...props,
  })

const validateExportTask = (task, srOrRemoteIds, props) => {
  expect(task).toMatchSnapshot({
    end: expect.any(Number),
    id: expect.any(String),
    message: expect.any(String),
    start: expect.any(Number),
    ...props,
  })
  expect(srOrRemoteIds).toContain(task.data.id)
}

const validateOperationTask = (task, props) => {
  expect(task).toMatchSnapshot({
    end: expect.any(Number),
    id: expect.any(String),
    message: expect.any(String),
    start: expect.any(Number),
    ...props,
  })
}

describe('backupNg', () => {
  let defaultBackupNg

  beforeAll(() => {
    defaultBackupNg = {
      name: 'default-backupNg',
      mode: 'full',
      vms: {
        id: config.vms.default,
      },
      settings: {
        '': {
          reportWhen: 'never',
        },
      },
    }
  })

  describe('.createJob() :', () => {
    it('creates a new backup job without schedules', async () => {
      const backupNg = await xo.createTempBackupNgJob(defaultBackupNg)
      expect(backupNg).toMatchSnapshot({
        id: expect.any(String),
        userId: expect.any(String),
        vms: expect.any(Object),
      })
      expect(backupNg.vms).toEqual(defaultBackupNg.vms)
      expect(backupNg.userId).toBe(xo._user.id)
    })

    it('creates a new backup job with schedules', async () => {
      const scheduleTempId = randomId()
      const { id: jobId } = await xo.createTempBackupNgJob({
        ...defaultBackupNg,
        schedules: {
          [scheduleTempId]: DEFAULT_SCHEDULE,
        },
        settings: {
          ...defaultBackupNg.settings,
          [scheduleTempId]: { snapshotRetention: 1 },
        },
      })

      const backupNgJob = await xo.call('backupNg.getJob', { id: jobId })

      expect(backupNgJob).toMatchSnapshot({
        id: expect.any(String),
        userId: expect.any(String),
        settings: expect.any(Object),
        vms: expect.any(Object),
      })
      expect(backupNgJob.vms).toEqual(defaultBackupNg.vms)
      expect(backupNgJob.userId).toBe(xo._user.id)

      expect(Object.keys(backupNgJob.settings).length).toBe(2)
      const schedule = await xo.getSchedule({ jobId })
      expect(typeof schedule).toBe('object')
      expect(backupNgJob.settings[schedule.id]).toEqual({
        snapshotRetention: 1,
      })

      expect(schedule).toMatchSnapshot({
        id: expect.any(String),
        jobId: expect.any(String),
      })
    })
  })

  describe('.delete() :', () => {
    it('deletes a backup job', async () => {
      const scheduleTempId = randomId()
      const { id: jobId } = await xo.call('backupNg.createJob', {
        ...defaultBackupNg,
        schedules: {
          [scheduleTempId]: DEFAULT_SCHEDULE,
        },
        settings: {
          ...defaultBackupNg.settings,
          [scheduleTempId]: { snapshotRetention: 1 },
        },
      })

      const schedule = await xo.getSchedule({ jobId })
      expect(typeof schedule).toBe('object')

      await xo.call('backupNg.deleteJob', { id: jobId })

      let isRejectedJobErrorValid = false
      await xo.call('backupNg.getJob', { id: jobId }).catch(error => {
        isRejectedJobErrorValid = noSuchObject.is(error)
      })
      expect(isRejectedJobErrorValid).toBe(true)

      let isRejectedScheduleErrorValid = false
      await xo.call('schedule.get', { id: schedule.id }).catch(error => {
        isRejectedScheduleErrorValid = noSuchObject.is(error)
      })
      expect(isRejectedScheduleErrorValid).toBe(true)
    })
  })

  describe('.runJob() :', () => {
    it('fails trying to run a backup job without schedule', async () => {
      const { id } = await xo.createTempBackupNgJob(defaultBackupNg)
      await expect(xo.call('backupNg.runJob', { id })).rejects.toMatchSnapshot()
    })

    it('fails trying to run a backup job with no matching VMs', async () => {
      const scheduleTempId = randomId()
      const { id: jobId } = await xo.createTempBackupNgJob({
        ...defaultBackupNg,
        schedules: {
          [scheduleTempId]: DEFAULT_SCHEDULE,
        },
        settings: {
          [scheduleTempId]: { snapshotRetention: 1 },
        },
        vms: {
          id: config.vms.default,
          name: 'test-vm-backupNg',
        },
      })

      const schedule = await xo.getSchedule({ jobId })
      expect(typeof schedule).toBe('object')

      await expect(
        xo.call('backupNg.runJob', { id: jobId, schedule: schedule.id })
      ).rejects.toMatchSnapshot()
    })

    it('fails trying to run a backup job with non-existent vm', async () => {
      jest.setTimeout(7e3)
      const scheduleTempId = randomId()
      const { id: jobId } = await xo.createTempBackupNgJob({
        ...defaultBackupNg,
        schedules: {
          [scheduleTempId]: DEFAULT_SCHEDULE,
        },
        settings: {
          [scheduleTempId]: { snapshotRetention: 1 },
        },
        vms: {
          id: 'non-existent-id',
        },
      })

      const schedule = await xo.getSchedule({ jobId })
      expect(typeof schedule).toBe('object')

      await xo.call('backupNg.runJob', { id: jobId, schedule: schedule.id })
      const [log] = await xo.call('backupNg.getLogs', {
        scheduleId: schedule.id,
      })
      expect(log.warnings).toMatchSnapshot()
    })

    it('fails trying to run a backup job with a VM without disks', async () => {
      jest.setTimeout(8e3)
      await xo.createTempServer(config.servers.default)
      const vmIdWithoutDisks = await xo.createTempVm({
        name_label: 'XO Test Without Disks',
        name_description: 'Creating a vm without disks',
        template: config.templates.default,
      })

      const scheduleTempId = randomId()
      const { id: jobId } = await xo.createTempBackupNgJob({
        ...defaultBackupNg,
        schedules: {
          [scheduleTempId]: DEFAULT_SCHEDULE,
        },
        settings: {
          ...defaultBackupNg.settings,
          [scheduleTempId]: { snapshotRetention: 1 },
        },
        vms: {
          id: vmIdWithoutDisks,
        },
      })

      const schedule = await xo.getSchedule({ jobId })
      expect(typeof schedule).toBe('object')
      await xo.call('backupNg.runJob', { id: jobId, schedule: schedule.id })

      const [
        {
          tasks: [vmTask],
          ...log
        },
      ] = await xo.call('backupNg.getLogs', {
        jobId,
        scheduleId: schedule.id,
      })
      expect(log).toMatchSnapshot({
        end: expect.any(Number),
        id: expect.any(String),
        jobId: expect.any(String),
        scheduleId: expect.any(String),
        start: expect.any(Number),
      })

      expect(vmTask).toMatchSnapshot({
        end: expect.any(Number),
        data: {
          id: expect.any(String),
        },
        id: expect.any(String),
        message: expect.any(String),
        result: {
          stack: expect.any(String),
        },
        start: expect.any(Number),
      })

      expect(vmTask.data.id).toBe(vmIdWithoutDisks)
    })

    it('fails trying to run backup job without retentions', async () => {
      jest.setTimeout(7e3)
      const scheduleTempId = randomId()
      await xo.createTempServer(config.servers.default)
      const { id: remoteId } = await xo.createTempRemote(config.remotes.default)
      const { id: jobId } = await xo.createTempBackupNgJob({
        ...defaultBackupNg,
        remotes: {
          id: remoteId,
        },
        schedules: {
          [scheduleTempId]: DEFAULT_SCHEDULE,
        },
        settings: {
          ...defaultBackupNg.settings,
          [scheduleTempId]: {},
        },
        srs: {
          id: config.srs.default,
        },
      })

      const schedule = await xo.getSchedule({ jobId })
      expect(typeof schedule).toBe('object')
      await xo.call('backupNg.runJob', { id: jobId, schedule: schedule.id })

      const [
        {
          tasks: [task],
          ...log
        },
      ] = await xo.call('backupNg.getLogs', {
        jobId,
        scheduleId: schedule.id,
      })

      expect(log).toMatchSnapshot({
        end: expect.any(Number),
        id: expect.any(String),
        jobId: expect.any(String),
        scheduleId: expect.any(String),
        start: expect.any(Number),
      })

      expect(task).toMatchSnapshot({
        data: {
          id: expect.any(String),
        },
        end: expect.any(Number),
        id: expect.any(String),
        message: expect.any(String),
        result: {
          stack: expect.any(String),
        },
        start: expect.any(Number),
      })
      expect(task.data.id).toBe(config.vms.default)
    })
  })

  test('execute three times a rolling snapshot with 2 as retention & revert to an old state', async () => {
    jest.setTimeout(6e4)
    await xo.createTempServer(config.servers.default)
    const vmId = await xo.createTempVm({
      name_label: 'XO Test Temp',
      name_description: 'Creating a temporary vm',
      template: config.templates.default,
      VDIs: [
        {
          size: 1,
          SR: config.srs.default,
          type: 'user',
        },
      ],
    })

    const scheduleTempId = randomId()
    const { id: jobId } = await xo.createTempBackupNgJob({
      ...defaultBackupNg,
      vms: {
        id: vmId,
      },
      schedules: {
        [scheduleTempId]: DEFAULT_SCHEDULE,
      },
      settings: {
        ...defaultBackupNg.settings,
        [scheduleTempId]: { snapshotRetention: 2 },
      },
    })

    const schedule = await xo.getSchedule({ jobId })
    expect(typeof schedule).toBe('object')
    for (let i = 0; i < 3; i++) {
      const oldSnapshots = xo.objects.all[vmId].snapshots
      await xo.call('backupNg.runJob', { id: jobId, schedule: schedule.id })
      await xo.waitObjectState(vmId, ({ snapshots }) => {
        // Test on updating snapshots.
        expect(snapshots).not.toEqual(oldSnapshots)
      })
    }

    const { snapshots, videoram: oldVideoram } = xo.objects.all[vmId]

    // Test on the retention, how many snapshots should be saved.
    expect(snapshots.length).toBe(2)

    const newVideoram = 16
    await xo.call('vm.set', { id: vmId, videoram: newVideoram })
    await xo.waitObjectState(vmId, ({ videoram }) => {
      expect(videoram).toBe(newVideoram.toString())
    })

    await xo.call('vm.revert', {
      snapshot: snapshots[0],
    })

    await xo.waitObjectState(vmId, ({ videoram }) => {
      expect(videoram).toBe(oldVideoram)
    })

    const [
      {
        tasks: [{ tasks: subTasks, ...vmTask }],
        ...log
      },
    ] = await xo.call('backupNg.getLogs', {
      jobId,
      scheduleId: schedule.id,
    })

    expect(log).toMatchSnapshot({
      end: expect.any(Number),
      id: expect.any(String),
      jobId: expect.any(String),
      scheduleId: expect.any(String),
      start: expect.any(Number),
    })

    const subTaskSnapshot = subTasks.find(
      ({ message }) => message === 'snapshot'
    )
    expect(subTaskSnapshot).toMatchSnapshot({
      end: expect.any(Number),
      id: expect.any(String),
      result: expect.any(String),
      start: expect.any(Number),
    })

    expect(vmTask).toMatchSnapshot({
      data: {
        id: expect.any(String),
      },
      end: expect.any(Number),
      id: expect.any(String),
      message: expect.any(String),
      start: expect.any(Number),
    })
    expect(vmTask.data.id).toBe(vmId)
  })

  test('execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval', async () => {
    jest.setTimeout(6e4)

    const { nfs, smb } = config.remotes
    const { id: nfsId } = await xo.createTempRemote(nfs)
    const remotes = [nfsId]
    if (smb !== undefined) {
      const { id: smbId } = await xo.createTempRemote(smb)
      remotes.push(smbId)
    }

    const scheduleTempId = randomId()
    const vmId = config.vms.vmToBackup
    const fullInterval = 2
    const { id: jobId } = await xo.createTempBackupNgJob({
      mode: 'delta',
      remotes: {
        id: {
          __or: remotes,
        },
      },
      schedules: {
        [scheduleTempId]: DEFAULT_SCHEDULE,
      },
      settings: {
        '': {
          reportWhen: 'never',
          fullInterval,
        },
        [nfs.id]: { deleteFirst: true },
        [scheduleTempId]: { exportRetention: 2 },
      },
      vms: {
        id: vmId,
      },
    })

    const schedule = await xo.getSchedule({ jobId })
    expect(typeof schedule).toBe('object')

    const nExecutions = 3
    const backupsByRemote = await xo.runBackupJob(jobId, schedule.id, {
      remotes,
      nExecutions,
    })
    forOwn(backupsByRemote, backups => expect(backups.length).toBe(2))

    const backupLogs = await xo.call('backupNg.getLogs', {
      jobId,
      scheduleId: schedule.id,
    })
    expect(backupLogs.length).toBe(nExecutions)

    backupLogs.forEach(({ tasks, ...log }, key) => {
      validateRootTask(log, {
        data: {
          mode: 'delta',
          reportWhen: 'never',
        },
        message: 'backup',
        status: 'success',
      })
      let vmTaskValidated = false
      tasks.forEach(({ tasks, ...vmTask }) => {
        if (vmTask.data !== undefined && vmTask.data.type === 'VM') {
          expect(vmTaskValidated).toBe(false)
          validateVmTask(vmTask, vmId, { status: 'success' })
          tasks.forEach(({ tasks, ...subTask }) => {
            if (subTask.message === 'snapshot') {
              validateSnapshotTask(subTask, { status: 'success' })
            }
            if (subTask.message === 'export') {
              validateExportTask(subTask, remotes, {
                data: {
                  id: expect.any(String),
                  isFull: key % fullInterval === 0,
                  type: 'remote',
                },
                status: 'success',
              })
              let mergeTaskKey, transferTaskKey
              tasks.forEach((operationTask, key) => {
                if (
                  operationTask.message === 'transfer' ||
                  operationTask.message === 'merge'
                ) {
                  validateOperationTask(operationTask, {
                    result: { size: expect.any(Number) },
                    status: 'success',
                  })
                  if (operationTask.message === 'transfer') {
                    mergeTaskKey = key
                  } else {
                    transferTaskKey = key
                  }
                }
              })
              expect(
                subTask.data.id === nfs.id
                  ? mergeTaskKey > transferTaskKey
                  : mergeTaskKey < transferTaskKey
              ).toBe(true)
            }
          })
          vmTaskValidated = true
        }
      })
    })
  })
})
