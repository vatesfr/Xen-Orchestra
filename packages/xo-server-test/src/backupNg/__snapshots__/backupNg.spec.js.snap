// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backupNg .createJob() : creates a new backup job with schedules 1`] = `
Object {
  "id": Any<String>,
  "mode": "full",
  "name": "default-backupNg",
  "settings": Any<Object>,
  "type": "backup",
  "userId": Any<String>,
  "vms": Any<Object>,
}
`;

exports[`backupNg .createJob() : creates a new backup job with schedules 2`] = `
Object {
  "cron": "0 * * * * *",
  "enabled": false,
  "id": Any<String>,
  "jobId": Any<String>,
  "name": "scheduleTest",
}
`;

exports[`backupNg .createJob() : creates a new backup job without schedules 1`] = `
Object {
  "id": Any<String>,
  "mode": "full",
  "name": "default-backupNg",
  "settings": Object {
    "": Object {
      "reportWhen": "never",
    },
  },
  "type": "backup",
  "userId": Any<String>,
  "vms": Any<Object>,
}
`;

exports[`backupNg .runJob() : fails trying to run a backup job with a VM without disks 1`] = `
Object {
  "data": Object {
    "mode": "full",
    "reportWhen": "never",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "jobId": Any<String>,
  "jobName": "default-backupNg",
  "message": "backup",
  "scheduleId": Any<String>,
  "start": Any<Number>,
  "status": "skipped",
}
`;

exports[`backupNg .runJob() : fails trying to run a backup job with a VM without disks 2`] = `
Object {
  "data": Object {
    "id": Any<String>,
    "type": "VM",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "result": Object {
    "message": "no disks found",
    "name": "Error",
    "stack": Any<String>,
  },
  "start": Any<Number>,
  "status": "skipped",
}
`;

exports[`backupNg .runJob() : fails trying to run a backup job with no matching VMs 1`] = `[JsonRpcError: unknown error from the peer]`;

exports[`backupNg .runJob() : fails trying to run a backup job with non-existent vm 1`] = `
Array [
  Object {
    "data": Object {
      "vms": Array [
        "non-existent-id",
      ],
    },
    "message": "missingVms",
  },
]
`;

exports[`backupNg .runJob() : fails trying to run a backup job without schedule 1`] = `[JsonRpcError: invalid parameters]`;

exports[`backupNg .runJob() : fails trying to run backup job without retentions 1`] = `
Object {
  "data": Object {
    "mode": "full",
    "reportWhen": "never",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "jobId": Any<String>,
  "jobName": "default-backupNg",
  "message": "backup",
  "scheduleId": Any<String>,
  "start": Any<Number>,
  "status": "failure",
}
`;

exports[`backupNg .runJob() : fails trying to run backup job without retentions 2`] = `
Object {
  "data": Object {
    "id": Any<String>,
    "type": "VM",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "result": Object {
    "message": "copy, export and snapshot retentions cannot both be 0",
    "name": "Error",
    "stack": Any<String>,
  },
  "start": Any<Number>,
  "status": "failure",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 1`] = `
Object {
  "data": Object {
    "mode": "delta",
    "reportWhen": "never",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "jobId": Any<String>,
  "message": "backup",
  "scheduleId": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 2`] = `
Object {
  "data": Object {
    "id": Any<String>,
    "type": "VM",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 3`] = `
Object {
  "end": Any<Number>,
  "id": Any<String>,
  "message": "snapshot",
  "result": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 4`] = `
Object {
  "data": Object {
    "id": Any<String>,
    "isFull": true,
    "type": "remote",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 5`] = `
Object {
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "result": Object {
    "size": Any<Number>,
  },
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 6`] = `
Object {
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "result": Object {
    "size": Any<Number>,
  },
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 7`] = `
Object {
  "data": Object {
    "mode": "delta",
    "reportWhen": "never",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "jobId": Any<String>,
  "message": "backup",
  "scheduleId": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 8`] = `
Object {
  "data": Object {
    "id": Any<String>,
    "type": "VM",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 9`] = `
Object {
  "end": Any<Number>,
  "id": Any<String>,
  "message": "snapshot",
  "result": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 10`] = `
Object {
  "data": Object {
    "id": Any<String>,
    "isFull": false,
    "type": "remote",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 11`] = `
Object {
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "result": Object {
    "size": Any<Number>,
  },
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 12`] = `
Object {
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "result": Object {
    "size": Any<Number>,
  },
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 13`] = `
Object {
  "data": Object {
    "mode": "delta",
    "reportWhen": "never",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "jobId": Any<String>,
  "message": "backup",
  "scheduleId": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 14`] = `
Object {
  "data": Object {
    "id": Any<String>,
    "type": "VM",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 15`] = `
Object {
  "end": Any<Number>,
  "id": Any<String>,
  "message": "snapshot",
  "result": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 16`] = `
Object {
  "data": Object {
    "id": Any<String>,
    "isFull": true,
    "type": "remote",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 17`] = `
Object {
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "result": Object {
    "size": Any<Number>,
  },
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a delta backup with 2 remotes, 2 as retention and 2 as fullInterval 18`] = `
Object {
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "result": Object {
    "size": Any<Number>,
  },
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a rolling snapshot with 2 as retention & revert to an old state 1`] = `
Object {
  "data": Object {
    "mode": "full",
    "reportWhen": "never",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "jobId": Any<String>,
  "jobName": "default-backupNg",
  "message": "backup",
  "scheduleId": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a rolling snapshot with 2 as retention & revert to an old state 2`] = `
Object {
  "end": Any<Number>,
  "id": Any<String>,
  "message": "snapshot",
  "result": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a rolling snapshot with 2 as retention & revert to an old state 3`] = `
Object {
  "data": Object {
    "id": Any<String>,
    "type": "VM",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;
